src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php

```diff
@@ -101,6 +142,91 @@ public function publish(string $body, array $headers = array()): void
         $this->exchange()->publish($body, null, AMQP_NOPARAM, array('headers' => $headers));
     }
 
+    /**
+     * @throws \AMQPException
+     */
+    public function publishForRetry(\AMQPEnvelope $message): bool
+    {
+        if (!isset($this->connectionConfiguration['retry'])) {
+            return false;
+        }
+
+        $retryConfiguration = $this->connectionConfiguration['retry'];
+        $attemptNumber = ((int) $message->getHeader('symfony-messenger-attempts') ?: 0) + 1;
+
+        if ($this->shouldSetup()) {
+            $this->setupRetry($retryConfiguration, $attemptNumber);
+        }
+
+        $maximumAttempts = $retryConfiguration['attempts'] ?? 3;
+        $routingKey = str_replace('%attempt%', $attemptNumber, $retryConfiguration['routing_key_pattern']);
+
+        if ($attemptNumber > $maximumAttempts) {
+            if (!isset($retryConfiguration['dead_queue'])) {
+                return false;
+            }
+
+            $routingKey = $retryConfiguration['dead_routing_key'];
+        }
+
+        $this->retryExchange($retryConfiguration)->publish(
+            $message->getBody(),
+            $routingKey,
+            AMQP_NOPARAM,
+            array('headers' => array_merge($message->getHeaders(), array('symfony-messenger-attempts' => (string) $attemptNumber)))
```

Comment:
a message has more attributes than just the headers like delivery_mode. do those attributes not get lost if you don't forward them?

Reply:
Good point, added the ones documented by the AMQP extension.